### PR TITLE
Misc test fixes for op serde

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Fix several tests including conditions, scatter add diff, softmax coss entropy loss
Focuses on test fixes relating to gradient checks.
Mostly java serialization related failures.
When we train a model, we create a gradient function which copies the graph
and appends a set of gradient ops for execution for our autodiff.
This graph created uses our flatbuffers support to copy graph configurations
from the base graph (the feed forward) to the training graph.

When we serialize models, various configurations and java op classes need to bootstrap configuration specific objects such as Conv2DConfig. This introduces several fixes related to test failures that were caused by serialization problems.

(Please fill in changes proposed in this fix)

## How was this patch tested?

Unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

### CODE REVIEW

File: libnd4j/include/ops/declarable/generic/loss/softmaxCrossEntropyWithLogits.cpp [hunk 1]
```diff
@@ -104,18 +104,28 @@ CUSTOM_OP_IMPL(softmax_cross_entropy_loss_with_logits_grad, 2, 2, false, 0, 0) {
...
+    auto labelsPlusEps = *labels + 1e-6;
+    labelsPlusEps.printBuffer("Labels plus eps");
+    dLdp->assign(((softmax * labelsPlusEps.reduceAlongDimension(reduce::Sum, dimension, true) - labelsPlusEps)));
+    auto negSoftmax = softmax;
+
+
+    // dEdl = -log(softmax)
+    softmax.applyTransform(transform::Log, *dLdl);
+    dLdl->applyTransform(transform::Neg,*dLdl);
+    return Status::OK();
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/loss/softmaxCrossEntropyWithLogits.cpp]: -> sd::Status:OK;

File: libnd4j/include/ops/declarable/generic/transforms/depth_to_space.cpp [hunk 2]
```diff
@@ -30,32 +30,68 @@
...
-  else
-    helpers::_depthToSpace(block.launchContext(), input->dup(), output, block_size, isNHWC);
+        if (shape::strideDescendingCAscendingF(input->shapeInfo()))
+            helpers::_depthToSpace(block.launchContext(), *input, output, block_size, isNHWC);
+        else
+            helpers::_depthToSpace(block.launchContext(), input->dup(), output, block_size, isNHWC);
+
+        STORE_RESULT(output);     
+
+        return ND4J_STATUS_OK;
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/transforms/depth_to_space.cpp]: sd::Status::OK;

File: libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp [hunk 3]
```diff
@@ -31,78 +31,72 @@ namespace sd {
...
+        ->setAllowedOutputTypes({ALL_INTS, ALL_FLOATS});
 }
 
-}  // namespace ops
-}  // namespace sd
+}
+}
 
-#endif
+#endif
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp:100]: need a new line at the end. A little context: some compilers warn such things which is annoying(nec nc++).  otherwise we can safely ignore it, https://stackoverflow.com/questions/72271/no-newline-at-end-of-file-compiler-warning

File: libnd4j/include/ops/declarable/generic/transforms/space_to_depth.cpp [hunk 4]
```diff
@@ -31,57 +31,69 @@
...
-  else
-    helpers::_spaceTodepth(block.launchContext(), input->dup(), output, block_size, isNHWC);
+        if (shape::strideDescendingCAscendingF(input->shapeInfo()))
+            helpers::_spaceTodepth(block.launchContext(), *input, output, block_size, isNHWC);
+        else
+            helpers::_spaceTodepth(block.launchContext(), input->dup(), output, block_size, isNHWC);
 
-  return sd::Status::OK;
-}
+        return Status::OK();
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/transforms/space_to_depth.cpp]: sd::Status::OK;

File: libnd4j/include/ops/declarable/generic/transforms/space_to_depth.cpp [hunk 5]
```diff
@@ -31,57 +31,69 @@
...
+        int iW = isNHWC ? shape::sizeAt(in, 2) : shape::sizeAt(in, 3);
 
-  int oD = iD * block_size * block_size;
-  int oH = iH / block_size;
-  int oW = iW / block_size;
+        int oD = iD * block_size * block_size;
+        int oH = iH / block_size;
+        int oW = iW / block_size;
+        
+        std::array<Nd4jLong, 4> shape;
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/transforms/space_to_depth.cpp]: sd::LongType

**Review by quickwritereader — CHANGES_REQUESTED** [review]
there are few things which breaks compilation. for formatting the libnd4j/.clang-format file should be used

**Review by quickwritereader — DISMISSED** [review]

File: libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp [hunk 6]
```diff
@@ -31,78 +31,72 @@ namespace sd {
...
+    if(inRank == 1) {
+        REQUIRE_TRUE(indices->isSameShape(updates), 0, "SCATTER_ADD OP: when input array has rank = 1 then indices and updates must have the same shapes, but got %s and %s correspondingly !", ShapeUtils::shapeAsString(indices).c_str(), ShapeUtils::shapeAsString(updates).c_str());
+    }
+    else if (inRank == updRank && indices->isVector()) {
+        std::vector<sd::LongType> updShape = updates->getShapeAsVector();
+        std::vector<sd::LongType> inShape  = input->getShapeAsVector();
+        std::vector<sd::LongType> expectedUpdShape = {indices->lengthOf()};
+        expectedUpdShape.insert(expectedUpdShape.end(), inShape.begin()+1, inShape.end());
+
+        //REQUIRE_TRUE(expectedUpdShape == updShape, 0, "SCATTER_ADD OP: wrong shape of updates array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedUpdShape).c_str(), ShapeUtils::shapeAsString(updShape).c_str());
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp]: Please don't leave any commented out code around. Either remove it or keep it, but just commenting things out without any indication for why that is, shouldn't make it into a merged PR.

File: libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp [hunk 7]
```diff
@@ -31,78 +31,72 @@ namespace sd {
...
 
-    helpers::scatter(block.launchContext(), pairwise::Add, *indices, *updates, *output, lock);
-  }
+        std::vector<sd::LongType> updShape = updates->getShapeAsVector();
+        std::vector<sd::LongType> inShape  = input->getShapeAsVector();
+        std::vector<sd::LongType> expectedUpdShape = indices->getShapeAsVector();
+        expectedUpdShape.insert(expectedUpdShape.end(), inShape.begin() + sd::LongType(1L), inShape.end());
 
-  return sd::Status::OK;
+        //REQUIRE_TRUE(expectedUpdShape == updShape, 0, "SCATTER_ADD OP: wrong shape of updates array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedUpdShape).c_str(), ShapeUtils::shapeAsString(updShape).c_str());
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/transforms/scatter_add.cpp]: Please don't leave any commented out code around

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/comparison/CompareAndSet.java [hunk 8]
```diff
@@ -45,13 +43,13 @@ public CompareAndSet(SameDiff sameDiff, SDVariable to, Number set, Condition con
...
         this.set = set.doubleValue();
-        this.mode = condition.condtionNum();
+        this.mode = condition.conditionNum();
         this.eps = condition.epsThreshold();
         this.extraArgs = new Object[] {compare, set, eps, (double) mode};
     }
 
     public CompareAndSet() {
-
+        System.out.println();
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/comparison/CompareAndSet.java]: was this left in after debugging?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java [hunk 9]
```diff
@@ -24,6 +24,29 @@
...
+        IS_INFINITE, //8
+        IS_NAN, //9
+        ABS_EQUALS, //10
+        NOT_EQUALS, //11
+        ABS_GREATER_OR_EQUAL, //12
+        ABS_LESS_THAN_OR_EQUAL, //13
+        IS_FINITE, //14
+        NOT_FINITE, //15
+
+    }
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: Would be nice to make that fixed so resorting those wouldn't result in changes. 

Also it would be very good to use them as Constants in all the places where we need to return a `conditionNum`

**Review by treo — CHANGES_REQUESTED** [review]
A few nit picks
